Update logging tool and control

borisdayma opened the pull request:
Following features have been implemented:
- logging separates pid for throttle and braking
- logging is enabled even if dbw is set to 0 (manual)
- logging was added for traffic light detection
- csv log files are now within specific instance directory "log/latest"
- there was an issue with the integral portion of PID (due to wrong update of intermediate value). In addition it was not capped and could grow high.
- steering is controlled using a lowpass filter. The issue with pid is that it is based on steering_error = target_steering - current_steering. However current_steering cannot be measured properly (we could only have used previous value). The Yaw Controller is supposed to already predict steering, which we make smooth by using the lowpass filter
- Feedforward was removed due to differences of calibration between Carla and simulator. Also due to the range of feedforward and pid, the command throttle could have discontinuous range.
- adjusted parameters for throttle PID based on trial and error (see below graph)
![image](https://user-images.githubusercontent.com/715491/34076785-d4faa350-e2b6-11e7-9724-38f423709ff2.png)
- set parameters for brake PID as percentages (due to undefined calibration on Carla vs simulator) and adjusted based on trial and error
![image](https://user-images.githubusercontent.com/715491/34077209-a5e53ae6-e2c3-11e7-9771-7826cc12f3b6.png)
- we added low pass filters to all commands, which create a smoother control, especially for braking.
![image](https://user-images.githubusercontent.com/715491/34077264-50348014-e2c5-11e7-9e8c-52588f6b6ed1.png)

borisdayma committed:
Update logging tool

Separate logging of throttle and brake pid's. Closes #38.
Ensure that logging is enabled regarless of dbw mode

borisdayma committed:
add logging of traffic light detection

Fixes #33

borisdayma committed:
csv log files now saved in "log/latest" and pid reset in manual mode

csv log files created by nodes are now saved in specific instance logging directory "log/latest"
PID's are now reset in manual mode. Fixes #27

borisdayma committed:
Correct integral portion of PID

There was a mistake in calculating the integral portion of the PID. Also it was replaced by an EMA to ensure it would not become too large.

borisdayma committed:
Use a low pass filter for controlling steering

Steering was controlled through PID but was unstable. Fixes #31.

borisdayma committed:
set up commands in percentage and adjust pid parameters

Command for brake control was set in percentage.
In addition, pid parameters were fine tuned. It closes #17, closes #18, and closes #28

borisdayma committed:
Added low pass filters for throttle and brake

Low pass filters have been added for throttle and brake to output smoother and continuous values. It closes #13.

borisdayma committed:
Update plotting script based on parameters logged

taikione commented:
I have confirmed this pr work correctly. thank you boris.

> logging is enabled even if dbw is set to 0 (manual)

Does this dbw mean `self.log_to_csv` on [dbw_node.py L72](https://github.com/borisd13/CarND-Capstone/pull/45/files#diff-7fdad75c966f68bdd51946534578bb1fR72)?

borisdayma committed:
Stop closer to stop lines

Adjusted target distance for stopping at red lights and decreased tau on low pass filters for better reactivity

borisd13 commented:
Yes that is correct Taiki. It also means [the change at L82](https://github.com/borisd13/CarND-Capstone/pull/45/files#diff-7fdad75c966f68bdd51946534578bb1fR82) ensures that the loop is executed even when dbw is False. The only value it does not get executed is when it is None (meaning it has not been initialized through self.upd_dbw_enabled yet).
